block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockMetadataStore.scala

```diff
@@ -0,0 +1,105 @@
+    for {
+      // Migrating data from previous metadata index
+//      result <- BlockMetadataPersistentIndex.read(Paths.get("index_path"))
+//      (oldMetadaatMap, _, _) = result
+//      _ <- blockMetadataStore.put(oldMetadaatMap.toSeq)
+      blockMetadataMap <- blockMetadataStore.toMap
+      dagState         = recreateInMemoryState(blockMetadataMap)
+    } yield new BlockMetadataStore[F](
+      blockMetadataStore,
+      new AtomicMonadState(AtomicAny(dagState))
+    )
+
+  private final case class DagState(
+      dagSet: Set[BlockHash],
+      childMap: Map[BlockHash, Set[BlockHash]],
+      heightMap: SortedMap[Long, Set[BlockHash]]
+  )
+
+  class BlockMetadataStore[F[_]: Monad](
+      private val store: KeyValueTypedStore[F, BlockHash, BlockMetadata],
+      private val dagState: MonadState[F, DagState]
+  ) {
+    def add(block: BlockMetadata): F[Unit] =
+      for {
+        // Update DAG state with new block
+        _ <- dagState.modify(st => validateDagState(addBlockToDagState(block)(st)))
+
+        // Update persistent block metadata store
+        _ <- store.put(block.blockHash, block)
+      } yield ()
+
+    def get(hash: BlockHash): F[Option[BlockMetadata]] = store.get(hash)
+
+    // DAG state operations
+
+    def dagSet: F[Set[BlockHash]] = dagState.get.map(_.dagSet)
+
+    def contains(hash: BlockHash): F[Boolean] = dagState.get.map(_.dagSet.contains(hash))
+
+    def childMapData: F[Map[BlockHash, Set[BlockHash]]] =
+      dagState.get.map(_.childMap)
+
+    def heightMap: F[SortedMap[Long, Set[BlockHash]]] =
+      dagState.get.map(_.heightMap)
+  }
+
+  private def addBlockToDagState(block: BlockMetadata)(state: DagState): DagState = {
+    // Update dag set / all blocks in the DAG
+    val newDagSet = state.dagSet + block.blockHash
+
+    // Update children relation map
+    val blockChilds = block.parents.map((_, Set(block.blockHash))) :+ (block.blockHash, Set())
+    val newChildMap = blockChilds.foldLeft(state.childMap) {
+      case (acc, (key, newChildren)) =>
+        val currChildren = acc.getOrElse(key, Set.empty[BlockHash])
+        acc.updated(key, currChildren ++ newChildren)
+    }
+
+    // Update block height map
+    val newHeightMap = if (!block.invalid) {
+      val currSet = state.heightMap.getOrElse(block.blockNum, Set())
+      state.heightMap.updated(block.blockNum, currSet + block.blockHash)
+    } else state.heightMap
+
+    state.copy(dagSet = newDagSet, childMap = newChildMap, heightMap = newHeightMap)
+  }
+
+  private def validateDagState(state: DagState): DagState = {
+    // Validate height map index (block numbers) are in sequence without holes
+    val m          = state.heightMap
+    val (min, max) = if (m.nonEmpty) (m.firstKey, m.lastKey + 1) else (0L, 0L)
+    assert(max - min == m.size.toLong, "DAG store height map has numbers not in sequence.")
+    state
+  }
+
+  private def recreateInMemoryState(
```

Comment:
@tgrospic hdy think, can we maintain DAG state directly in LMDB without in-memory proxy?

Reply:
@nzpr this in-memory part gives as ability like a snapshots of the DAG. With LMDB is not that easy, we are always seeing the latest state of the data.
Now we have combination of LMDB to store the whole dag data and in-memory part is just the keys to filter the "heavy" part in LMDB.
So performance is not the concern here and moving more data to LMDB should be way to go. Also with LMDB allocation of direct buffer memory will increase which must be taken into account.